spec.html

```diff
@@ -22528,9 +22528,9 @@ <h1>
         1. If _R_ is an abrupt completion, return ? UpdateEmpty(_R_, _V_).
         1. NOTE: The following is another complete iteration of the second |CaseClauses|.
         1. For each |CaseClause| _C_ of _B_, do
-          1. Let _R_ be Completion(Evaluation of |CaseClause| _C_).
-          1. If _R_.[[Value]] is not ~empty~, set _V_ to _R_.[[Value]].
-          1. If _R_ is an abrupt completion, return ? UpdateEmpty(_R_, _V_).
+          1. Let _R2_ be Completion(Evaluation of |CaseClause| _C_).
+          1. If _R2_.[[Value]] is not ~empty~, set _V_ to _R2_.[[Value]].
+          1. If _R2_ is an abrupt completion, return ? UpdateEmpty(_R2_, _V_).
```

Comment:
In the status quo, this algorithm has 4 occurrences of `Let _R_ be ...`, so it's unclear why you think only one of them should be changed.

Reply:
This is the only part of the algorithm where the `R` variable is shadows an outer declaration of a variable with the same identifier, the other declarations are within scopes that do not have an outer declaration with the same identifier, before this `for each`, the `R` variable is declared at the global scope, and the `R` variable in this `for each`'s scope shadows it.

![Picsart_23-08-19_20-05-40-541.jpg](https://github.com/tc39/ecma262/assets/62040526/8f2e6470-b6d1-484b-85d0-a798c1b52634)

Green: declared within a scope where no previous **outer** declarations with the same identifier exist.
Red: declared within a scope where an **outer** declaration with the same identifier exists.
Blue: the declaration that is shadowed by the red declaration.

Reply:
Note that these were also reported by Ecmarkup's linter:

```console
> ecma262@1.0.0 build-only
> ecmarkup --verbose spec.html --multipage out --lint-spec --strict

[2023-08-19T13:57:04.046Z] Loading biblios...
[2023-08-19T13:57:04.151Z] Loading imports...
[2023-08-19T13:57:04.790Z] Building boilerplate...
[2023-08-19T13:57:05.223Z] Linting...
[2023-08-19T13:57:14.306Z] Error: spec.html:22531:18: "R" is already declared
[2023-08-19T13:57:15.210Z] Error: spec.html:31193:95: "f" is already declared
```

Reply:
After reading through a bit more, do these inner scopes not actually exist? Like are all of these declarations done at the global scope?

Reply:
It's a bit tricky. The spec certainly doesn't talk about algorithm aliases being scoped. If it were going to, the place to do so would be in [5.2 Algorithm Conventions](https://tc39.es/ecma262/#sec-algorithm-conventions), but all it says is:
> Once declared, an alias may be referenced in any subsequent steps and must not be referenced from steps prior to the alias's declaration. Aliases may be modified using the form “Set _x_ to _someOtherValue_”.

So that suggests that, roughly speaking, there's only one scope for the aliases in an algorithm.

Interestingly, the spec has lots of things like:
```
1. If <some condition>, then
   a. Let _x_ be ...
2. Else,
   a. Let _x_ be ...
3. Use _x_.
```
which tells you a couple things:
- The body of an `If` step is definitely not a scope, otherwise `_x_` wouldn't escape to step 3.
- The presence of multiple `Let _x_ be ...` in an algorithm isn't *necessarily* a problem, despite what Algorithm Conventions says. I think the intended rule is something more like: you shouldn't have a `Let _x_` at a point where `_x_` is already defined. (However, that isn't the rule that ecmarkup is using, otherwise it would complain more in this algorithm.)

Reply:
Ah I see, that actually helps a lot to understand them.

However I have a question, in the second change I made (when I changed `_f_` to `_g_`), I noticed the following step:

```txt
Let n, k, and s be integers such that k ≥ 1, radixk - 1 ≤ s < radixk, 𝔽(s × radixn - k) is x, and k is as small as possible. Note that k is the number of digits in the representation of s using radix radix, that s is not divisible by radix, and that the least significant digit of s is not necessarily uniquely determined by these criteria.
```

How does this work? What values do the `n`, `k`, and `s` variables get? Or if they don't then what does this do?
Just trying to learn so I can be more familiar and make better contributions.

Reply:
> ... in the second change I made (when I changed `_f_` to `_g_`), I noticed the following step:
> 
> > Let n, k, and s be integers such that k ≥ 1, radixk - 1 ≤ s < radixk, 𝔽(s × radixn - k) is x, and k is as small as possible. Note that k is the number of digits in the representation of s using radix radix, that s is not divisible by radix, and that the least significant digit of s is not necessarily uniquely determined by these criteria.
> 
> How does this work?

I'm a bit puzzled. Are you saying that you understand how the step of your second change works, but you don't understand how the above step works? They work about the same: find 3 integers that satisfy the constraints.

> What values do the `n`, `k`, and `s` algorithm aliases get?

The integer values that satisfy the constraints.

> Or if they don't then what does this do?

If a situation could arise in which it were impossible to find values to satisfy the constraints, the step would have a sentence something like: "If no such values exist, then...". The lack of such a sentence means that the spec authors are confident that the constraints can always be satisfied. It shouldn't be too hard to find a proof to that effect.

Reply:
> Are you saying that you understand how the step of your second change works, but you don't understand how the above step works?

I don't understand how either of those steps work but based on your response, there are constraints sure, but from where do they find these values in a way that it satisfied the constraints? Like are they supposed to use something declared in previous steps to find these values? Or how is that done?

Reply:
> ... from where do they find these values in a way that it satisfies the constraints?

The implementation is free to find these values in any way it likes. Normally it will try to find them efficiently.

> Like are they supposed to use something declared in previous steps to find these values?

Such constraints refer to one or more aliases that are defined in previous steps, so the values of those aliases will necessarily be used in the solution of the constraints. Is that what you mean?

Reply:
Hmm alright so the algorithm step I showed is actually from [`Number::toString()` § 6.1.6.1.20](https://tc39.es/ecma262/#sec-numeric-types-number-tostring) at step 5, what value does this step even use here to perform any kind of calculation to find values it needs (that satisfies the constraints)? I don't see how an implementation would perform these calculations without knowing which value they're supposed to use that is defined in the previous steps, can you show an example implementation and what value `n`, `k`, and `s` would get given an example `x` value (for example, `5`)?

Reply:
> what value does this step even use here to perform any kind of calculation to find the values it needs (that satisfies the constraints)?

It uses the values of _x_ and _radix_.

> I don't see how an implementation would perform these calculations without knowing which value they're supposed to use that is defined in the previous steps,

Why do you think the implementation wouldn't know?

> can you show an example implementation and what value `n`, `k`, and `s` would get given an example `x` value (for example, `5`)?

If _radix_ is 10, then _k_ is 1, _s_ is 5, and _n_ is 1.
Constraints:
- _k_ ≥ 1, i.e. 1 ≥ 1, true
- _radix_^(_k_ - 1) ≤ _s_ < _radix_^(_k_), i.e. 1 ≤ 5 < 10, true
- 𝔽(_s_ × _radix_^(_n_ - _k_)) is _x_, i.e. 𝔽(5 × 10^0)) is 5, true
- _k_ is as small as possible. Well, it can't be any smaller and still satisfy _k_ ≥ 1, so true.

If _radix_ is 2, then _k_ is 3, _s_ is 5, and _n_ is 3.
- _k_ ≥ 1, i.e. 3 ≥ 1, true
- _radix_^(_k_ - 1) ≤ _s_ < _radix_^(_k_), i.e. 4 ≤ 5 < 8, true
- 𝔽(_s_ × _radix_^(_n_ - _k_)) is _x_, i.e. 𝔽(5 × 2^0)) is 5, true
- _k_ is as small as possible. If _k_ were 1 or 2, then the only values for _s_ that satisfy the 2nd constraint (1, 2, or 3) can't also satisfy the third constraint.

Reply:
I see, that helped me fully understand how such steps work, thank you.

> Why do you think the implementation wouldn't know?

Oh I was speaking from my perspective, because I didn't realize what values were used in such steps, so I thought how would the implementations know but now I see.

Thank you so much for your time.